content/courses/advanced-ada/chapters/types.rst

```diff
@@ -1227,14 +1227,167 @@ Record Representation and storage clauses
 Valid
 -----
 
-.. admonition:: Relevant topics
+When receiving data from external sources, we're subjected to problems such as
+transmission errors. If not handled properly, erroneous data can lead to major
+issues in an application.
 
-    - `The Valid Attribute <http://www.ada-auth.org/standards/2xrm/html/RM-13-9-2.html>`_
+One of those issues originates from the fact that transmission errors might
+lead to invalid information stored in memory. When proper checks are activate,
+using invalid information is detected at runtime and an exception is raised at
+this point, which might then be handled by the application.
 
-.. todo::
+Instead of relying on exception handling, however, we could instead ensure that
+the information we're about to use is valid. We can do this by using the
+:ada:`Valid` attribute. For example, if we have a variable :ada:`Var`, we can
+verify that the value stored in :ada:`Var` is valid by writing
+:ada:`Var'Valid`, which returns a :ada:`Boolean` value. Therefore, if the value
+of :ada:`Var` isn't valid, :ada:`Var'Valid` returns :ada:`False`, so we can
+have code that handles this situation before we actually make use of
+:ada:`Var`. In other words, instead of handling a potential exception in other
+parts of the application, we can proactively verify that input information is
+correct and avoid that an exception is raised.
 
-    Complete section!
+In the next example, we show an application that
+
+- generates a file containing mock-up data, and then
+
+- reads information from this file as state values.
+
+The mock-up data includes valid and invalid states.
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Types.Valid_States
+    :class: ada-run
+
+    procedure Create_Test_File (File_Name : String);
+
+    with Ada.Sequential_IO;
+
+    procedure Create_Test_File (File_Name : String) is
+       package Integer_Sequential_IO is new Ada.Sequential_IO (Integer);
+       use Integer_Sequential_IO;
+
+       F : File_Type;
+    begin
+       Create (F, Out_File, File_Name);
+       Write (F,  0);
+       Write (F,  1);
+       Write (F,  2);
+       Write (F,  10);
+       Write (F,  1);
+       Write (F,  20);
+       Close (F);
+    end Create_Test_File;
+
+    with Ada.Sequential_IO;
+
+    package States is
+
+       type State is (Off, On, Waiting)
+         with Size => Integer'Size;
+
+       for State use (Off => 0, On => 1, Waiting => 2);
```

Comment:
Those are the default guaranteed representational values, so do you want to mention that this is a confirming clause?

Reply:
You're absolutely right! My original intention was to use values that are not consecutive... I've just changed that here (and in the #584).